### PR TITLE
feat: garm runners will be reflected into kubernetes as runner resource

### PR DESCRIPTION
**Pull request by rafalgalaw**
with 
```bash
$ kubectl get runner
```
one can now list and with 

```bash
$ kubectl delete runner my-garm-runner
```
one can delete a runner from within a cluster.
Also, if accidently a runner get deleted via cli from within the garm-server itself, der controller cleans up not matching runner CRs in k8s

Todo:
- [x] implement tests

### CODE REVIEW

File: api/v1alpha1/runner_types.go [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+// RunnerSpec defines the desired state of Runner
+type RunnerSpec struct {
+	ID string `json:"id,omitempty"`
+}
+
+// RunnerStatus defines the observed state of Runner
+type RunnerStatus struct {
+	// ID is the database ID of this instance.
+	ID string `json:"id,omitempty"`
```
Comment (bavarianbidi) [inline api/v1alpha1/runner_types.go:22]:
```suggestion
```

can we get rid of the `ID` in the status if the `ID` is already in the spec?

File: api/v1alpha1/runner_types.go [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+	// PeoviderID is the unique ID the provider associated
```
Comment (bavarianbidi) [inline api/v1alpha1/runner_types.go:24]:
```suggestion
	// ProviderID is the unique ID the provider associated
```

File: api/v1alpha1/runner_types.go [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+	// ProviderFault holds any error messages captured from the IaaS provider that is
+	// responsible for managing the lifecycle of the runner.
+	ProviderFault []byte `json:"providerFault,omitempty"`
+
+	// StatusMessages is a list of status messages sent back by the runner as it sets itself
+	// up.
+
+	//// UpdatedAt is the timestamp of the last update to this runner.
+	// UpdatedAt time.Time `json:"updated_at"`
+
```
Comment (bavarianbidi) [inline api/v1alpha1/runner_types.go:73]:
```suggestion
```

File: api/v1alpha1/runner_types.go [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+	Addresses []commonParams.Address `json:"addresses,omitempty"`
+
+	// Status is the status of the instance inside the provider (eg: running, stopped, etc)
+	Status commonParams.InstanceStatus `json:"status,omitempty"`
+
+	// RunnerStatus is the github runner status as it appears on GitHub.
+	InstanceStatus params.RunnerStatus `json:"instanceStatus,omitempty"`
+
+	// PoolID is the ID of the garm pool to which a runner belongs.
+	PoolID string `json:"poolId,omitempty"`
```
Comment (bavarianbidi) [inline api/v1alpha1/runner_types.go:62]: can we also get the pool Name (name of the corresponding CR) - as handling the pool IDs from a user perspective might be super ugly :sweat_smile:

File: api/v1alpha1/runner_types.go [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+	// GithubRunnerGroup is the github runner group to which the runner belongs.
+	// The runner group must be created by someone with access to the enterprise.
+	GitHubRunnerGroup string `json:"githubRunnerGroup"`
+}
+
+//+kubebuilder:object:root=true
+//+kubebuilder:resource:path=runners,scope=Namespaced,categories=garm,shortName=run
+//+kubebuilder:subresource:status
+//+kubebuilder:printcolumn:name="ID",type="string",JSONPath=".status.id",description="Runner ID"
+//+kubebuilder:printcolumn:name="Pool ID",type="string",JSONPath=".status.poolId",description="Pool ID"
```
Comment (bavarianbidi) [inline api/v1alpha1/runner_types.go:83]: would like to prefer the pool name instead of the ID (if possible)

File: api/v1alpha1/runner_types.go [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+//+kubebuilder:printcolumn:name="Garm Runner Status",type="string",JSONPath=".status.status",description="Garm Runner Status"
+//+kubebuilder:printcolumn:name="Provider Runner Status",type="string",JSONPath=".status.instanceStatus",description="Provider Runner Status"
+
```
Comment (bavarianbidi) [inline api/v1alpha1/runner_types.go:86]: With a priority of `1` (`//+kubebuilder:printcolumn:name=[...],priority=1`) we can thing about adding few more information to the output, e.g. `AgentID` and `ProviderID`)

File: config/rbac/role.yaml [hunk 7]
```diff
@@ -136,3 +136,29 @@ rules:
   - get
   - patch
   - update
+- apiGroups:
+  - garm-operator.mercedes-benz.com
+  resources:
+  - runners
+  verbs:
+  - create
```
Comment (bavarianbidi) [inline config/rbac/role.yaml:144]: should we really allow the creation of `runners` as the controller doesn't implemented this yet

File: config/rbac/runner_editor_role.yaml [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+    app.kubernetes.io/part-of: garm-operator
+    app.kubernetes.io/managed-by: kustomize
+  name: runner-editor-role
+rules:
+- apiGroups:
+  - garm-operator.mercedes-benz.com
+  resources:
+  - runners
+  verbs:
+  - create
```
Comment (bavarianbidi) [inline config/rbac/runner_editor_role.yaml:19]: same as above - creation isn't implemented

File: internal/controller/runner_controller.go [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+	client.Client
+	Scheme   *runtime.Scheme
+	Recorder record.EventRecorder
+
+	BaseURL  string
+	Username string
+	Password string
+}
+
+//+kubebuilder:rbac:groups=garm-operator.mercedes-benz.com,namespace=xxxxx,resources=runners,verbs=get;list;watch;create;update;patch;delete
```
Comment (bavarianbidi) [inline internal/controller/runner_controller.go:41]: as in the generated `yaml` already mentioned, we should remote the `create` param here as we do not create runners

```suggestion
//+kubebuilder:rbac:groups=garm-operator.mercedes-benz.com,namespace=xxxxx,resources=runners,verbs=get;list;watch;update;patch;delete
```

File: internal/controller/runner_controller.go [hunk 10]
```diff
@@ -0,0 +1,275 @@
...
+	instanceClient := garmClient.NewInstanceClient()
+	err = instanceClient.Login(garmClient.GarmScopeParams{
+		BaseURL:  r.BaseURL,
+		Username: r.Username,
+		Password: r.Password,
+		// Debug:    true,
+	})
+	if err != nil {
+		return err
+	}
```
Comment (bavarianbidi) [inline internal/controller/runner_controller.go:245]: We've intentionally decided to not "cache" the JWT token as we do not requesting the API-Server that often ([ADR](https://github.com/mercedes-benz/garm-operator/blob/main/docs/adrs/jwt_token_caching.md)).

But with this implementation i'm really struggling with that decision.
For this 5 second ticker we could think about the following for now:
* doing the login in `PollRunnerInstances` 
* passing the token to `EnqueueRunnerInstance`
  * If the JWT has expired, we might get a `401` and could catch that error
  * re-login and updating the JWT

At least we do not have to to the entire Login flow every 5 seconds

**Review by bavarianbidi — CHANGES_REQUESTED** [review]
Looks very nice - just few nits

:pray: please squash your commits so we will get nice autogenerated release notes. Otherwise each commit will appear in the release notes as one topic in the feature list

File: config/rbac/role.yaml (hunk 7 above)
Comment (rafalgalaw) [reply config/rbac/role.yaml:144]: we probably need create permissions, otherwise the operator might not be able to create runner CR with its service account... but creating runner CR manually is not much of a big deal anyway, as the controller will check if the CR has a corresponding entry in the garm DB, and if not will just clean it up.

Comment (bavarianbidi) [reply config/rbac/role.yaml:144]: thanks for the explanation

File: cmd/main.go [hunk 11]
```diff
@@ -97,86 +106,116 @@ func main() {
...
+		Username: config.Config.Garm.Username,
+		Password: config.Config.Garm.Password,
+		Debug:    false,
+	})
+
+	err = baseClient.Login()
+	if err != nil {
+		setupLog.Error(err, "unable to Login into GARM Client")
+		exitCode = 1
+		return
```
Comment (bavarianbidi) [inline cmd/main.go:124]: shouldn't it be enough to just exit here (`os.exit(1)`) if we are not allowed to login into garm - and with that, kubernetes will re-start the pod again (with the backoff-interval).

the current behavior of the operator is: that each reconciliation-loop is totally independent and if one reconciler crashes, the other controller loops are still continuing.
With your proposed implementation, an error in any of the reconciliation loops will immediately stop the entire program.

@rafalgalaw what was the intention for changing the current behavior?

**Comment by rafalgalaw** [comment]
After communication with @bavarianbidi, I will move the auth refactoring to a separate PR

File: docs/adrs/jwt_token_caching.md
Comment (bavarianbidi) [inline docs/adrs/jwt_token_caching.md:1]: we should revert this change as well and introduce the updated ADR together with the planned refactoring

**Comment by rafalgalaw** [comment]
- [ ] make tests work again after refactoring

File: docs/config/configuration-parsing.md [hunk 12]
```diff
@@ -45,6 +45,7 @@ OPERATOR_HEALTH_PROBE_BIND_ADDRESS
 OPERATOR_LEADER_ELECTION
 OPERATOR_SYNC_PERIOD
 OPERATOR_WATCH_NAMESPACE
+OPERATOR_DISABLE_WEBHOOKS
```
Comment (bavarianbidi) [inline docs/config/configuration-parsing.md:48]:
```suggestion
```

File: docs/config/configuration-parsing.md [hunk 13]
```diff
@@ -92,6 +94,7 @@ operator:
   leaderelection: false
   syncperiod: 5m0s
   watchnamespace: garm-operator-system
+  disablewebhooks: false
```
Comment (bavarianbidi) [inline docs/config/configuration-parsing.md:97]:
```suggestion
```

File: docs/config/configuration-parsing.md [hunk 14]
```diff
@@ -113,6 +116,7 @@ operator:
   leader_election: true
   sync_period: "10m"
   watch_namespace: "garm-operator-namespace"
+  disable_webhooks: false
```
Comment (bavarianbidi) [inline docs/config/configuration-parsing.md:119]:
```suggestion
```

File: docs/adrs/jwt_token_caching.md [hunk 15]
```diff
@@ -1,9 +1,9 @@
 <!-- SPDX-License-Identifier: MIT -->
 
 ---
-date: 2023-07-24
+date: 2023-12-04
```
Comment (bavarianbidi) [inline docs/adrs/jwt_token_caching.md:4]: as discussed - remove this ADR record as well and bring it back once the PR with the Auth refactoring is ready to review

File: internal/controller/runner_controller.go [hunk 16]
```diff
@@ -0,0 +1,354 @@
...
+
+	return c.Watch(&source.Channel{Source: eventChan}, &handler.EnqueueRequestForObject{})
+}
+
+func (r *RunnerReconciler) PollRunnerInstances(ctx context.Context, eventChan chan event.GenericEvent) {
+	log := log.FromContext(ctx)
+	ticker := time.NewTicker(5 * time.Second)
```
Comment (bavarianbidi) [inline internal/controller/runner_controller.go:203]: should we have this configurable? maybe it's a nice feature - wdyt?

File: internal/controller/runner_controller.go [hunk 17]
```diff
@@ -0,0 +1,354 @@
...
+	runnerCRNameList := slices.Map(runnerCRList.Items, func(runner garmoperatorv1alpha1.Runner) string {
+		return runner.Name
+	})
+
+	runnerInstanceNameList := slices.Map(garmRunnerInstances, func(runner params.Instance) string {
+		return strings.ToLower(runner.Name)
+	})
+
+	runnersToDelete := getRunnerDiff(runnerCRNameList, runnerInstanceNameList)
+	fmt.Println(runnersToDelete)
```
Comment (bavarianbidi) [inline internal/controller/runner_controller.go:285]: this should be a log message - (i would have this as a debug-message `log.V(1).Info`)

File: internal/controller/runner_controller.go [hunk 18]
```diff
@@ -0,0 +1,354 @@
...
+	runner.Status.Name = garmRunner.Name
+	runner.Status.OSType = garmRunner.OSType
+	runner.Status.OSName = garmRunner.OSName
+	runner.Status.OSVersion = garmRunner.OSVersion
+	runner.Status.OSArch = garmRunner.OSArch
+	runner.Status.Addresses = garmRunner.Addresses
+	runner.Status.Status = garmRunner.Status
+	runner.Status.InstanceStatus = garmRunner.RunnerStatus
+	runner.Status.PoolID = poolName
+	runner.Status.ProviderFault = garmRunner.ProviderFault
```
Comment (bavarianbidi) [inline internal/controller/runner_controller.go:179]: can we convert the byte-array to a string (to get more readable error messages):

this is what i currently get

```yaml
apiVersion: garm-operator.mercedes-benz.com/v1alpha1
kind: Runner
metadata:
  creationTimestamp: "2023-12-08T12:58:05Z"
  finalizers:
  - garm-operator.mercedes-benz.com/runner
  generation: 1
  name: garm-xn2dldt5ae7r
  namespace: garm-operator-system
  resourceVersion: "35165"
  uid: a2b34960-7fd9-425a-9eb9-064ffa7252fc
spec: {}
status:
  agentId: 0
  githubRunnerGroup: ""
  id: 01cc1bd6-e0e8-45f2-ab16-82a8d82aebde
  instanceStatus: pending
  name: garm-xN2dldT5aE7R
  osArch: amd64
  osType: linux
  poolId: k8s-pool
  providerFault: Y3JlYXRpbmcgaW5zdGFuY2U6IHByb3ZpZGVyIGJpbmFyeSAvb3B0L2dhcm0vYmluL2dhcm0tcHJvdmlkZXItazhzIHJldHVybmVkIGVycm9yOiBwcm92aWRlciBiaW5hcnkgZmFpbGVkIHdpdGggc3Rkb3V0OiA7IHN0ZGVycjogMjAyMy8xMi8wOCAxMjo1ODo1OCBmYWlsZWQgdG8gcnVuIGNvbW1hbmQ6IGZhaWxlZCB0byBjcmVhdGUgaW5zdGFuY2UgaW4gcHJvdmlkZXI6IGVuc3VyaW5nIHJ1bm5lciBuYW1lc3BhY2UgcnVubmVyIGZhaWxlZDogbmFtZXNwYWNlcyAicnVubmVyIiBpcyBmb3JiaWRkZW46IFVzZXIgInN5c3RlbTpzZXJ2aWNlYWNjb3VudDpnYXJtLXNlcnZlcjpnYXJtLXNlcnZlciIgY2Fubm90IGdldCByZXNvdXJjZSAibmFtZXNwYWNlcyIgaW4gQVBJIGdyb3VwICIiIGluIHRoZSBuYW1lc3BhY2UgInJ1bm5lciIKOiBleGl0IHN0YXR1cyAx
  status: error
```

which means:

```
echo "Y3JlYXRpbmcgaW5zdGFuY2U6IHByb3ZpZGVyIGJpbmFyeSAvb3B0L2dhcm0vYmluL2dhcm0tcHJvdmlkZXItazhzIHJldHVybmVkIGVycm9yOiBwcm92aWRlciBiaW5hcnkgZmFpbGVkIHdpdGggc3Rkb3V0OiA7IHN0ZGVycjogMjAyMy8xMi8wOCAxMjo1ODozOCBmYWlsZWQgdG8gcnVuIGNvbW1hbmQ6IGZhaWxlZCB0byBjcmVhdGUgaW5zdGFuY2UgaW4gcHJvdmlkZXI6IGVuc3VyaW5nIHJ1bm5lciBuYW1lc3BhY2UgcnVubmVyIGZhaWxlZDogbmFtZXNwYWNlcyAicnVubmVyIiBpcyBmb3JiaWRkZW46IFVzZXIgInN5c3RlbhcnkgZmFpbGVkIHdpdGggc3Rkb3V0OiTpzZXJ2aWNlYWNjb3VudDpnYXJtLXNlcnZlcjpnYXJtLXNlcnZlciIgY2Fubm90IGdldCByZXNvdXJjZSAibmFtZXNwYWNlcyIgaW4gQVBJIGdyb3VwICIiIGluIHRoZSBuYW1lHJ1bm5lciBuYW1lc3BhY2UgcnVubmVyc3BhY2UgInJ1bm5lciIKOiBleGl0IHN0YXR1cyAx" | base64 -d                                                                                  0IGdldCByZXNvdXJjZSAibmFtZXNwYW
creating instance: provider binary /opt/garm/bin/garm-provider-k8s returned error: provider binary failed with stdout: ; stderr: 2023/12/08 12:58:38 failed to run command: failed to create instance in provider: ensuring runner namespace runner failed: namespaces "runner2/08 12:58:38 failed to run com" is forbidden: User "system:serviceaccount:garm-server:garm-server" cannot get resource "namespaces" in API group "" in the namespace rviceaccount:garm-server:garm-s"runner"
: exit status 1
```

File: internal/controller/runner_controller.go (hunk 16 above)
Comment (bavarianbidi) [reply internal/controller/runner_controller.go:206]: as discussed, i'm fine with adding an additional flag in a separate PR
